### PR TITLE
Handle MILC staggerd phase removal in QUDA

### PR DESCRIPTION
**Pull request by mathiaswagner**
This is the first steps towards better gauge field residency in MILC. 
It adds functions to the MILC interface that take an argument `phase_in` that signals QUDA that MILC has not removed the staggered phases and QUDA can handle it.
This removes the need for MILC to modify the gauge field in the HMC evolution.

This does not break compatibility with older MILC version but modifications to MILC are required to exploit the benefits. Those can be found in https://github.com/lattice/milc_qcd/tree/feature/quda_rephase

I'll file a PR for that later.

### CODE REVIEW

File: include/quda_milc_interface.h [hunk 1]
```diff
@@ -662,6 +662,25 @@ extern "C" {
...
+   * @param eb3 The integration step size (for MILC this is dt*beta/3)
+   * @param arg Metadata for MILC's internal site struct array
+   * @param phase_in whether staggered phases are applied
+   */
+  void qudaGaugeForcePhased(int precision,
+		      int num_loop_types,
+		      double milc_loop_coeff[3],
+		      double eb3,
+		      QudaMILCSiteArg_t *arg,
+          int phase_in);
```
Comment (weinbe2) [inline include/quda_milc_interface.h]: Needs a `clang-format`

File: lib/interface_quda.cpp [hunk 2]
```diff
@@ -4932,6 +4934,7 @@ void updateGaugeFieldQuda(void* gauge,
    if (param->use_resident_gauge) {
      if (!gaugePrecise) errorQuda("No resident gauge field to use");
      cudaGauge = gaugePrecise;
+     gaugePrecise = nullptr;
```
Comment (weinbe2) [inline lib/interface_quda.cpp]: Is this addition safe? I thought `gaugePrecise` could incidentally get used by other routines. Whether or not that happens in practice though MILC is another story.

File: lib/milc_interface.cpp [hunk 3]
```diff
@@ -538,6 +554,11 @@ void qudaGaugeForce( int precision,
   qudaGaugeParam.mom_offset = arg->mom_offset;
   qudaGaugeParam.site_size = arg->size;
   qudaGaugeParam.gauge_order = arg->site ? QUDA_MILC_SITE_GAUGE_ORDER : QUDA_MILC_GAUGE_ORDER;
+  qudaGaugeParam.staggered_phase_applied = phase_in;
+  qudaGaugeParam.staggered_phase_type = QUDA_STAGGERED_PHASE_MILC;
+  if(phase_in) qudaGaugeParam.t_boundary    = QUDA_ANTI_PERIODIC_T;
```
Comment (weinbe2) [inline lib/milc_interface.cpp]: Can we always assume MILC is using antiperiodic bc during RHMC?

File: lib/milc_interface.cpp [hunk 4]
```diff
@@ -1223,10 +1253,10 @@ void qudaCloverForce(void *mom, double dt, void **x, void **p, double *coeff, do
 		     int nvec, double multiplicity, void *gauge, int precision, QudaInvertArgs_t inv_args)
 {
   qudamilc_called<true>(__func__);
-  QudaGaugeParam gaugeParam = newMILCGaugeParam(localDim,
+  QudaGaugeParam qudaGaugeParam = newMILCGaugeParam(localDim,
 						(precision==1) ? QUDA_SINGLE_PRECISION : QUDA_DOUBLE_PRECISION,
 						QUDA_GENERAL_LINKS);
-  gaugeParam.gauge_order = QUDA_MILC_GAUGE_ORDER; // refers to momentume gauge order
+  qudaGaugeParam.gauge_order = QUDA_MILC_GAUGE_ORDER; // refers to momentume gauge order
```
Comment (weinbe2) [inline lib/milc_interface.cpp]: Incidentally noticed: typo in comment

File: include/quda_milc_interface.h (hunk 1 above)
Comment (mathiaswagner) [reply include/quda_milc_interface.h]: indeed.

File: lib/milc_interface.cpp [hunk 5]
```diff
@@ -318,67 +318,83 @@ void qudaComputeOprod(int prec, int num_terms, int num_naik_terms, double** coef
...
-  gaugeParam.site_size = arg->size;
-  gaugeParam.gauge_order = arg->site ? QUDA_MILC_SITE_GAUGE_ORDER : QUDA_MILC_GAUGE_ORDER;
-
-  projectSU3Quda(gauge, tol, &gaugeParam);
+  qudaGaugeParam.gauge_offset = arg->link_offset;
+  qudaGaugeParam.site_size = arg->size;
+  qudaGaugeParam.gauge_order = arg->site ? QUDA_MILC_SITE_GAUGE_ORDER : QUDA_MILC_GAUGE_ORDER;
+  qudaGaugeParam.staggered_phase_applied = phase_in;
+  qudaGaugeParam.staggered_phase_type = QUDA_STAGGERED_PHASE_MILC;
+  qudaGaugeParam.t_boundary    = QUDA_ANTI_PERIODIC_T;
```
Comment (weinbe2) [inline lib/milc_interface.cpp]: Same comment as below, can we always assume this?

**Comment by weinbe2** [comment]
I assume this has been tested with the MILC PR so I'm not worried about RHMC correctness via that benchmark, but I'm worried some of the additions related to boundary conditions could be breaking in the general sense.

**Comment by mathiaswagner** [comment]
Thanks for your comments. I'll leave them open. We can maybe remove some of the args. I had just decided to split the work on a resident gauge field for MILC HMC into two PR to make them easier to digest. Some changes might have slipped in here that are not (yet) necessary.

File: include/quda_milc_interface.h (hunk 1 above)
Comment (mathiaswagner) [reply include/quda_milc_interface.h]: done

File: lib/milc_interface.cpp (hunk 3 above)
Comment (mathiaswagner) [reply lib/milc_interface.cpp]: removed. Not necessary here.

File: lib/milc_interface.cpp (hunk 5 above)
Comment (mathiaswagner) [reply lib/milc_interface.cpp]: Interpreting the comments from Carleton yes.

File: lib/milc_interface.cpp (hunk 4 above)
Comment (mathiaswagner) [reply lib/milc_interface.cpp]: fixed

File: lib/interface_quda.cpp (hunk 2 above)
Comment (mathiaswagner) [reply lib/interface_quda.cpp]: removed.

File: lib/milc_interface.cpp (hunk 3 above)
Comment (mathiaswagner) [reply lib/milc_interface.cpp]: Interpreting the comments from Carleton yes.

**Review by maddyscientist — APPROVED** [review]
Good solution to the lack of default parameters with C.

**Review by weinbe2 — APPROVED** [review]
Visual code review is good, I believe I tested it too but it's been a few weeks. This is good contingent on Jenkins build tests and @maddyscientist testing it against MILC (in progress right now).

**Comment by detar** [comment]
I have merged the milc_qcd PR #36 and fixed a slew of side effects in 
the MILC code.  The latest develop branch has the new code.  What branch 
of QUDA has merged QUDA PR #1018?

**Comment by maddyscientist** [comment]
This has now been verified against the MILC NERSC benchmarks.  The fixes from @detar's end were what was needed at the MILC end to get correctness for these

**Pull request merged**